samza-core/src/main/java/org/apache/samza/table/remote/RemoteReadableTable.java

```diff
@@ -113,51 +106,53 @@ public void init(SamzaContainerContext containerContext, TaskContext taskContext
   @Override
   public V get(K key) {
     try {
-      readMetrics.numGets.inc();
-      if (rateLimitReads) {
-        throttle(key, null, RL_READ_TAG, readCreditFn, getThrottleNs);
-      }
-      long startNs = System.nanoTime();
-      V result = readFn.get(key);
-      readMetrics.getNs.update(System.nanoTime() - startNs);
-      return result;
+      return getAsync(key).get();
     } catch (Exception e) {
-      String errMsg = String.format("Failed to get a record, key=%s", key);
-      logger.error(errMsg, e);
-      throw new SamzaException(errMsg, e);
+      throw new SamzaException(e);
     }
   }
 
+  @Override
+  public CompletableFuture<V> getAsync(K key) {
+    Preconditions.checkNotNull(key);
+    readMetrics.numGets.inc();
+    return asyncHelper.execute(key, readFn::getAsync, readMetrics.getNs)
+        .exceptionally(e -> {
+            throw new SamzaException("Failed to get the record for " + key, e);
+          });
+  }
+
   /**
    * {@inheritDoc}
    */
   @Override
   public Map<K, V> getAll(List<K> keys) {
-    Map<K, V> result;
+    readMetrics.numGetAlls.inc();
     try {
-      readMetrics.numGetAlls.inc();
-      long startNs = System.nanoTime();
-      result = readFn.getAll(keys);
-      readMetrics.getAllNs.update(System.nanoTime() - startNs);
+      return getAllAsync(keys).get();
     } catch (Exception e) {
-      String errMsg = "Failed to get some records";
-      logger.error(errMsg, e);
-      throw new SamzaException(errMsg, e);
-    }
-
-    if (result == null) {
-      String errMsg = String.format("Received null records, keys=%s", keys);
-      logger.error(errMsg);
-      throw new SamzaException(errMsg);
+      throw new SamzaException(e);
     }
+  }
 
-    if (result.size() < keys.size()) {
-      String errMsg = String.format("Received insufficient number of records (%d), keys=%s", result.size(), keys);
-      logger.error(errMsg);
-      throw new SamzaException(errMsg);
+  @Override
+  public CompletableFuture<Map<K, V>> getAllAsync(List<K> keys) {
+    Preconditions.checkNotNull(keys);
+    if (keys.isEmpty()) {
+      return CompletableFuture.completedFuture(Collections.EMPTY_MAP);
     }
-
-    return result;
+    readMetrics.numGetAlls.inc();
+    return asyncHelper.execute(keys, readFn::getAllAsync, readMetrics.getAllNs)
+        .handle((result, e) -> {
+            if (e != null) {
+              throw new SamzaException("Failed to get the records for " + keys, e);
+            } else {
+              if (result != null && result.size() != keys.size()) {
```

Comment:
It can happen when the semantics of the native client for batchGet is to return whatever results it has for the keys, which can be partial when some of the keys do not exist in the store, right? I know DataVault ACL service (restli) does this and only return existing ACLs for the set of keys. 

But I can see when getAll is used in low-level, it might not be necessary for us to enforce such rules when the app expects partial results, then this would forbid that kind of use cases. 

Based on this, I think maybe it makes sense to just let partial results pass. I can relax such rule.

Reply:
Yes, let it pass seem to make more sense. At the end of the day, user can always check and we are not claiming the full set.